File: Lista 5/Python/Contato.py
from datetime import datetime as dt
from Data import Data

class Contato:
    def __new__(cls, *args, **kwargs):
        return super().__new__(cls)
    
    def __init__(self, nome, email, telefone, data_nascimento):
        self.__validarNome(nome)
        self.__validarEmail(email)
        self.__nome = nome
        self.__email = email
        self.__telefone = telefone
        self.__data_nascimento = data_nascimento

    # Imprimir contato
    def __str__(self):
        return f"Nome: {self.__nome}\n" \
               f"Email: {self.__email}\n" \
               f"Telefone: {self.__telefone}\n" \
               f"Data de Nascimento: {self.__data_nascimento}"
    
    def __validarNome(self, nome: str):
        if not (nome == "" and nome.isspace()):
            return True
        raise ValueError("Invalid name")
    
    def __validarEmail(self, email: str):
        if email.endswith(".com") or email.endswith(".br"):
            return True
        raise ValueError("Invalid email")

    # Calcular idade
    def calcular_idade(self):
        diferenca = dt.today() - dt.strptime(str(self.__data_nascimento), "%d/%m/%Y")
        idade = int(diferenca.days / 365.25)
        return idade

    # Getters
    @property
    def nome(self):
        return self.__nome
    
    @property
    def email(self):
        return self.__email
    
    @property
    def telefone(self):
        return self.__telefone
    
    @property
    def data_nascimento(self):
        return self.__data_nascimento

    # Setters
    @email.setter
    def email(self, email):
        self.__email = email

    @telefone.setter
    def telefone(self, telefone):
        self.__telefone = telefone

if __name__ == "__main__":
    dataNascimento = Data(9, 5, 2000)
    contato = Contato("Fulano", "fulano@email.com", "73999999999", dataNascimento)
    print(contato)
    print("Idade", contato.calcular_idade())

File: Lista 7/Java/Contato.java
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

public class Contato {
    private String nome;
    private String email;
    private String telefone;
    private LocalDate dataNascimento;
    private static int numContatos = 0;

    Contato(String nome, String email, String telefone, LocalDate dataNascimento){
        this.nome = nome;
        this.email = email;
        this.telefone = telefone;
        this.dataNascimento = dataNascimento;
        numContatos++;
    }

    Contato(String nome, String telefone, LocalDate dataNascimento){
        this(nome, "Nao informado", telefone, dataNascimento);
    }

    // Imprimir contato
    public void imprimirContato(){
        System.out.println("Nome: " + nome);
        System.out.println("Email: " + email);
        System.out.println("Telefone: " + telefone);
        System.out.println("Data de Nascimento: " + DateTimeFormatter.ofPattern("dd/MM/yyyy").format(dataNascimento));
    }

    // Calcular idade
    public int calcularIdade(){
        LocalDate dataHoje = LocalDate.now();
        int idade = dataHoje.getYear() - dataNascimento.getYear();
        if (dataHoje.getMonthValue() < dataNascimento.getMonthValue()
                || (dataHoje.getMonthValue() == dataNascimento.getMonthValue() && dataHoje.getDayOfMonth() < dataNascimento.getDayOfMonth()))
            idade--;

        return idade;
    }

    public static void printNumContatos() {
        System.out.println("Numero de contatos: " + numContatos);
    }

    // Getters
    public String getNome(){
        return nome;
    }

    public String getEmail(){
        return email;
    }

    public String getTelefone(){
        return telefone;
    }
    
    public LocalDate getDataNascimento(){
        return dataNascimento;
    }

    // Setters
    public void setEmail(String email){
        this.email = email;
    }

    public void setTelefone(String telefone){
        this.telefone = telefone;
    }

    public static void main(String[] args){
        LocalDate data = LocalDate.of(2023, 5, 8);
        Contato contato1 = new Contato("Fulano", "73999999999", data);
        Contato contato2 = new Contato("Ciclano", "73999999999", data);
        Contato contato3 = new Contato("Beltrano", "73999999999", data);
        Contato.printNumContatos();
    }
    
}
